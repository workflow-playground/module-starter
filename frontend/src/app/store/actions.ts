import { actions as books } from './slices/books.store';
import { actions as collection } from './slices/collection.store';
import { actions as search } from './slices/search.store';
import { actions as auth } from './slices/auth.store';
import { actions as layout } from './slices/layout.store';
import { actions as loginPage } from './slices/login-page.store';
import { actions as user } from './slices/user.store';

export { books, collection, search, auth, loginPage, layout, user };
