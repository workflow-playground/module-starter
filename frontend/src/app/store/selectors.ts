import { selectors as books } from './slices/books.store';
import { selectors as collection } from './slices/collection.store';
import { selectors as search } from './slices/search.store';
import { selectors as layout } from './slices/layout.store';
import { selectors as auth } from './slices/auth.store';
import { selectors as loginPage } from './slices/login-page.store';
import { selectors as router } from './slices/router.store';

export {
  router,
  books,
  collection,
  search,
  layout,
  auth,
  loginPage
}
