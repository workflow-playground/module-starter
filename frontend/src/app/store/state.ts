import { RouterReducerState } from '@ngrx/router-store';

import * as books from './slices/books.store';
import * as collection from './slices/collection.store';
import * as search from './slices/search.store';
import * as auth from './slices/auth.store';
import * as loginPage from './slices/login-page.store';
import * as layout from './slices/layout.store';

/**
 * As mentioned, we treat each reducer like a table in a database. This means
 * our top level state interface is just a map of keys to inner state types.
 */
export interface State {
  [books.featureKey]: books.State;
  [collection.featureKey]: collection.State;
  [search.featureKey]: search.State;
  [auth.featureKey]: auth.State;
  [loginPage.featureKey]: loginPage.State;
  [layout.featureKey]: layout.State;
  router: RouterReducerState;
}
