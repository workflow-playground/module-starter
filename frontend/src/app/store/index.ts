import * as selectors from './selectors';
import * as actions from './actions';
import { State } from './state';

export {
  selectors,
  State,
  actions
}
