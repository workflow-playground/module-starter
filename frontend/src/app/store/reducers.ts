import { isDevMode } from '@angular/core';
import { MetaReducer, ActionReducerMap } from '@ngrx/store';
import { routerReducer } from '@ngrx/router-store';

import { State } from './state';

/**
 * Every reducer module's default export is the reducer function itself. In
 * addition, each module should export a type or interface that describes
 * the state of the reducer plus any selector functions. The `* as`
 * notation packages up all of the exports into a single object.
 */

import * as books from './slices/books.store';
import * as collection from './slices/collection.store';
import * as search from './slices/search.store';
import * as auth from './slices/auth.store';
import * as loginPage from './slices/login-page.store';
import * as layout from './slices/layout.store';

/**
 * Our state is composed of a map of action reducer functions.
 * These reducer functions are called with each dispatched action
 * and the current or initial state and return a new immutable state.
 */
export const rootReducers: ActionReducerMap<State> = {
  [books.featureKey]: books.reducer,
  [collection.featureKey]: collection.reducer,
  [search.featureKey]: search.reducer,
  [auth.featureKey]: auth.reducer,
  [loginPage.featureKey]: loginPage.reducer,
  [layout.featureKey]: layout.reducer,
  router: routerReducer,
};

/**
 * By default, @ngrx/store uses combineReducers with the reducer map to compose
 * the root meta-reducer. To add more meta-reducers, provide an array of meta-reducers
 * that will be composed to form the root meta-reducer.
 */
export const metaReducers: MetaReducer<State>[] = isDevMode() ? [] : [];
