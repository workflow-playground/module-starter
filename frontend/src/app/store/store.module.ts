import { NgModule, isDevMode } from '@angular/core';
import { StoreModule } from '@ngrx/store';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { rootReducers, metaReducers } from './reducers';

@NgModule({
  imports: [
    /**
     * StoreModule.forRoot is imported once in the root module, accepting a reducer
     * function or object map of reducer functions. If passed an object of
     * reducers, combineReducers will be run creating your application
     * meta-reducer. This returns all providers for an @ngrx/store
     * based application.
     */
    StoreModule.forRoot(rootReducers, {
      metaReducers,
      runtimeChecks: {
        // strictStateImmutability and strictActionImmutability are enabled by default
        strictStateSerializability: true,
        strictActionSerializability: true,
        strictActionWithinNgZone: true,
        strictActionTypeUniqueness: true,
      },
    }),

    /**
     * @ngrx/router-store keeps router state up-to-date in the store.
     */
    StoreRouterConnectingModule.forRoot(),

    /**
     * Store devtools instrument the store retaining past versions of state
     * and recalculating new states. This enables powerful time-travel
     * debugging.
     *
     * To use the debugger, install the Redux Devtools extension for either
     * Chrome or Firefox
     *
     * See: https://github.com/zalmoxisus/redux-devtools-extension
     */
    StoreDevtoolsModule.instrument({
      name: 'NgRx Book Store App',
      // In a production build you would want to disable the Store Devtools
      logOnly: !isDevMode(),
    }),
  ],
})
export class AppStoreModule {}
