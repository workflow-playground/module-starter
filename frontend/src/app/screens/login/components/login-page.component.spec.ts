import { TestBed, ComponentFixture } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MemoizedSelector } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { TranslateModule } from '@ngx-translate/core';

import { LoginPageComponent } from './login-page.component';
import * as selectors from '../../../store/selectors';
import { State } from '../../../store/state';
import * as actions from '../../../store/actions';

describe('Login Page', () => {
  let fixture: ComponentFixture<LoginPageComponent>;
  let instance: LoginPageComponent;
  let mockStore: MockStore<State>;
  let selectLoginPageError: MemoizedSelector<State, string | null>;
  let selectLoginPagePending: MemoizedSelector<State, boolean>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [ReactiveFormsModule, TranslateModule.forRoot()],
      declarations: [LoginPageComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [provideMockStore()],
    });

    fixture = TestBed.createComponent(LoginPageComponent);
    instance = fixture.componentInstance;
    mockStore = TestBed.inject(MockStore);
    selectLoginPageError = mockStore.overrideSelector(selectors.loginPage.selectLoginPageError, null);
    selectLoginPagePending = mockStore.overrideSelector(selectors.loginPage.selectLoginPagePending, false);

    jest.spyOn(mockStore, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should disable the form if pending', () => {
    selectLoginPagePending.setResult(true);
    mockStore.refreshState();
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should display an error message if provided', () => {
    selectLoginPageError.setResult('Invalid Credentials');
    mockStore.refreshState();
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch an action if the form is valid when submitted', () => {
    const credentials = {
      username: 'user',
      password: 'pass',
    };

    instance.form.setValue(credentials);


    instance.submit();

    expect(mockStore.dispatch).toHaveBeenCalledWith(actions.loginPage.login({ credentials }));
  });
});
