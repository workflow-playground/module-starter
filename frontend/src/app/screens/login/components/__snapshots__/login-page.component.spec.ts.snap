// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should compile 1`] = `
<ds-login-page
  _document={[Function Document]}
  error$={[Function _Store]}
  form={[Function FormGroup]}
  ngAfterContentChecked$={[Function Subject]}
  ngAfterContentInit$={[Function Subject]}
  ngAfterViewChecked$={[Function Subject]}
  ngAfterViewInit$={[Function Subject]}
  ngDoCheck$={[Function Subject]}
  ngOnChanges$={[Function Subject]}
  ngOnDestroy$={[Function Subject]}
  ngOnInit$={[Function Subject]}
  store={[Function _MockStore]}
>
  <div
    class="flex min-h-full flex-1 flex-col justify-center px-6 py-12 lg:px-8"
  >
    <div
      class="sm:mx-auto sm:w-full sm:max-w-sm"
    >
      <h2
        class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
      >
        Sign in to your account
      </h2>
    </div>
    <div
      class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
    >
      <form
        class="space-y-6 ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <ds-input-wrapper>
          <input
            class="ng-untouched ng-pristine ng-valid"
            dsinput=""
            formcontrolname="username"
            placeholder="LOGIN.USERNAME.PLACEHOLDER"
            type="text"
          />
          <ds-form-label>
            LOGIN.USERNAME.LABEL
          </ds-form-label>
        </ds-input-wrapper>
        <ds-input-wrapper>
          <input
            class="ng-untouched ng-pristine ng-valid"
            dsinput=""
            formcontrolname="password"
            type="password"
          />
          <ds-form-label>
            LOGIN.PASSWORD.LABEL
          </ds-form-label>
          <ds-form-sub-text>
            LOGIN.PASSWORD.HINT
          </ds-form-sub-text>
        </ds-input-wrapper>
        <div>
          <button
            class="flex w-full justify-center rounded-md"
            dsbutton="primary"
            type="submit"
          >
            LOGIN.SIGN_IN
          </button>
        </div>
        <div
          class="w-full text-center"
        >
        </div>
      </form>
    </div>
  </div>
</ds-login-page>
`;

exports[`Login Page should disable the form if pending 1`] = `
<ds-login-page
  _document={[Function Document]}
  error$={[Function _Store]}
  form={[Function FormGroup]}
  ngAfterContentChecked$={[Function Subject]}
  ngAfterContentInit$={[Function Subject]}
  ngAfterViewChecked$={[Function Subject]}
  ngAfterViewInit$={[Function Subject]}
  ngDoCheck$={[Function Subject]}
  ngOnChanges$={[Function Subject]}
  ngOnDestroy$={[Function Subject]}
  ngOnInit$={[Function Subject]}
  store={[Function _MockStore]}
>
  <div
    class="flex min-h-full flex-1 flex-col justify-center px-6 py-12 lg:px-8"
  >
    <div
      class="sm:mx-auto sm:w-full sm:max-w-sm"
    >
      <h2
        class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
      >
        Sign in to your account
      </h2>
    </div>
    <div
      class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
    >
      <form
        class="space-y-6 ng-untouched ng-pristine"
        novalidate=""
      >
        <ds-input-wrapper>
          <input
            class="ng-untouched ng-pristine"
            disabled=""
            dsinput=""
            formcontrolname="username"
            placeholder="LOGIN.USERNAME.PLACEHOLDER"
            type="text"
          />
          <ds-form-label>
            LOGIN.USERNAME.LABEL
          </ds-form-label>
        </ds-input-wrapper>
        <ds-input-wrapper>
          <input
            class="ng-untouched ng-pristine"
            disabled=""
            dsinput=""
            formcontrolname="password"
            type="password"
          />
          <ds-form-label>
            LOGIN.PASSWORD.LABEL
          </ds-form-label>
          <ds-form-sub-text>
            LOGIN.PASSWORD.HINT
          </ds-form-sub-text>
        </ds-input-wrapper>
        <div>
          <button
            class="flex w-full justify-center rounded-md"
            dsbutton="primary"
            type="submit"
          >
            LOGIN.SIGN_IN
          </button>
        </div>
        <div
          class="w-full text-center"
        >
        </div>
      </form>
    </div>
  </div>
</ds-login-page>
`;

exports[`Login Page should display an error message if provided 1`] = `
<ds-login-page
  _document={[Function Document]}
  error$={[Function _Store]}
  form={[Function FormGroup]}
  ngAfterContentChecked$={[Function Subject]}
  ngAfterContentInit$={[Function Subject]}
  ngAfterViewChecked$={[Function Subject]}
  ngAfterViewInit$={[Function Subject]}
  ngDoCheck$={[Function Subject]}
  ngOnChanges$={[Function Subject]}
  ngOnDestroy$={[Function Subject]}
  ngOnInit$={[Function Subject]}
  store={[Function _MockStore]}
>
  <div
    class="flex min-h-full flex-1 flex-col justify-center px-6 py-12 lg:px-8"
  >
    <div
      class="sm:mx-auto sm:w-full sm:max-w-sm"
    >
      <h2
        class="mt-10 text-center text-2xl font-bold leading-9 tracking-tight text-gray-900"
      >
        Sign in to your account
      </h2>
    </div>
    <div
      class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
    >
      <form
        class="space-y-6 ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <ds-input-wrapper>
          <input
            class="ng-untouched ng-pristine ng-valid"
            dsinput=""
            formcontrolname="username"
            placeholder="LOGIN.USERNAME.PLACEHOLDER"
            type="text"
          />
          <ds-form-label>
            LOGIN.USERNAME.LABEL
          </ds-form-label>
        </ds-input-wrapper>
        <ds-input-wrapper>
          <input
            class="ng-untouched ng-pristine ng-valid"
            dsinput=""
            formcontrolname="password"
            type="password"
          />
          <ds-form-label>
            LOGIN.PASSWORD.LABEL
          </ds-form-label>
          <ds-form-sub-text>
            LOGIN.PASSWORD.HINT
          </ds-form-sub-text>
        </ds-input-wrapper>
        <div>
          <button
            class="flex w-full justify-center rounded-md"
            dsbutton="primary"
            type="submit"
          >
            LOGIN.SIGN_IN
          </button>
        </div>
        <div
          class="w-full text-center"
        >
          <ds-form-error>
            Invalid Credentials
          </ds-form-error>
        </div>
      </form>
    </div>
  </div>
</ds-login-page>
`;
